tools/build_defs/repo/local.bzl

```diff
@@ -0,0 +1,130 @@
+# .bzl files in this package (tools/build_defs/repo) are evaluated
+# in a Starlark environment without "@_builtins" injection, and must not refer
+# to symbols associated with build/workspace .bzl files
+
+"""Rules for making directories in the local filesystem available as repos.
+
+### Setup
+
+To use these rules in a module extension, load them in your .bzl file and then call them from your
+extension's implementation function. For example, to use `local_repository`:
+
+```python
+load("@bazel_tools//tools/build_defs/repo:local.bzl", "local_repository")
+
+def _my_extension_impl(mctx):
+  local_repository(name = "foo", path = "foo")
+
+my_extension = module_extension(implementation = _my_extension_impl)
+```
+
+Alternatively, you can directly call these repo rules in your MODULE.bazel file with
+`use_repo_rule`:
+
+```python
+local_repository = use_repo_rule("@bazel_tools//tools/build_defs/repo:local.bzl", "local_repository")
+local_repository(name = "foo", path = "foo")
+```
+"""
+
+# The default value for the string attr `build_file_content`. String attrs default to the empty
+# string by default, but we need another default because the empty string is perfectly valid build
+# file content and we need to know whether the attribute is actually set.
+_UNSET = "_UNSET"
+
+def _get_dir_path(rctx):
+    """Turns the string attr `path` into a path object, ensuring that it's a directory."""
+    path = rctx.workspace_root.get_child(rctx.attr.path)
+    if not path.is_dir:
+        fail(
+            ("The repository's path is \"%s\" (absolute: \"%s\") but it does not exist or is not " +
+             "a directory.") % (rctx.attr.path, path),
+        )
+    return path
+
+def _local_repository_impl(rctx):
+    rctx.symlink(_get_dir_path(rctx), ".")
+
+local_repository = repository_rule(
+    implementation = _local_repository_impl,
+    attrs = {
+        "path": attr.string(
+            doc =
+                "The path to the directory to make available as a repo. <p>The path can be " +
+                "either absolute, or relative to the workspace root.",
+            mandatory = True,
+        ),
+    },
+    doc =
+        "Makes a local directory that already contains Bazel files available as a repo. This " +
+        "directory should contain Bazel BUILD files and a repo boundary file already. If it " +
+        "doesn't contain these files, consider using <a " +
+        "href=\"#new_local_repository\"><code>new_local_repository</code></a> instead.",
+    local = True,
+)
+
+def _new_local_repository_impl(rctx):
+    if (rctx.attr.build_file == None) == (rctx.attr.build_file_content == _UNSET):
+        fail("exactly one of `build_file` and `build_file_content` must be specified")
+
+    children = _get_dir_path(rctx).readdir()
+    for child in children:
+        rctx.symlink(child, child.basename)
+
+        # On Windows, `rctx.symlink` actually does a copy. So we need to watch the symlink target as
```

Comment:
Afaik directories can always be symlinked on Windows (as junctions), so maybe we only need to watch files? If that's not true, wouldn't we have to watch directories recursively?

We could also check the OS and skip the additional watch if it isn't needed.

Reply:
you're absolutely right -- amended to only watch files & on windows